salt/states/keystone.py

```diff
@@ -703,48 +698,58 @@ def endpoint_present(name,
 
             if endpoint.get('publicurl', None) != publicurl:
                 change_publicurl = True
+                ret['comment'] = _changes('Public URL will be updated: {0} -> {1}'.format(endpoint.get('publicurl', None), publicurl))
 
             if endpoint.get('adminurl', None) != adminurl:
                 change_adminurl = True
+                ret['comment'] = _changes('Admin URL will be updated: {0} -> {1}'.format(endpoint.get('adminurl', None), adminurl))
 
             if endpoint.get('internalurl', None) != internalurl:
                 change_internalurl = True
+                ret['comment'] = _changes('Internal URL will be updated: {0} -> {1}'.format(endpoint.get('internal', None), internal))
 
             if __opts__.get('test') and (change_publicurl or change_adminurl or change_internalurl):
                 ret['result'] = None
-                ret['comment'] = 'Endpoint for service "{0}" will be updated'.format(name)
+                ret['comment'] += 'Endpoint for service "{0}" will be updated'.format(name)
                 ret['changes']['Endpoint'] = 'Will be updated'
                 return ret
 
             if endpoint.get('publicurl', None) != publicurl:
-                ret['comment'] = 'Endpoint for service "{0}" will be updated'.format(name)
+                ret['comment'] += 'Endpoint for service "{0}" will be updated'.format(name)
                 ret['changes']['publicurl'] = publicurl
 
             if endpoint.get('adminurl', None) != adminurl:
-                ret['comment'] = 'Endpoint for service "{0}" will be updated'.format(name)
+                ret['comment'] += 'Endpoint for service "{0}" will be updated'.format(name)
                 ret['changes']['adminurl'] = adminurl
 
             if endpoint.get('internalurl', None) != internalurl:
-                ret['comment'] = 'Endpoint for service "{0}" will be updated'.format(name)
+                ret['comment'] += 'Endpoint for service "{0}" will be updated'.format(name)
                 ret['changes']['internalurl'] = internalurl
 
-        if ret['comment'].endswith('already exists'):
+        if ret['comment'] == '':
+            ret['comment'] = 'Endpoint for service "{0}" already exists'.format(name)}
             return ret
         else:
-            __salt__['keystone.endpoint_delete'](name,
+            # Delete endpoint
+            __salt__['keystone.endpoint_delete'](name, region,
                                                  profile=profile,
                                                  **connection_args)
-            ret['comment'] = 'Endpoint for service "{0}" has been updated'.format(name)
-            ret['changes']['Endpoint'] = 'Will be updated'
+            ret['comment'] = _changes('Endpoint has been deleted and re-created')
+            # Re create
+            _addEndpoint()
```

Comment:
If I am in the picture, we also do not test creation of the endpoints over api v3 in the unit tests.